### PR TITLE
Add trace to help debug IPC message corruption in runner.

### PR DESCRIPTION
**Pull request by TingluoHuang**
Help better debug https://github.com/github/c2c-actions-support/issues/650

TLDR, job message JSON over IPC somehow corrupted, the received string length is equal to what is sent, but the actual content has changed.

With this trace, we should be able to figure out what is corrupted over the IPC between runner and worker.

Example trace:
```
Runner sent Json at position '9272': "TingluoHuang"},{"k" (IlRpbmdsdW9IdWFuZyJ9LHsiayI=)
Worker received invalid Json at position '9272': "Tingluo"}Huang"},{" (IlRpbmdsdW8ifUh1YW5nIn0seyI=)
Newtonsoft.Json.JsonReaderException: After parsing a value an unexpected character was encountered: H. Path 'contextData.github.d[9]', line 1, position 9272.
```

### CODE REVIEW

File: src/Runner.Listener/JobDispatcher.cs [hunk 1]
```diff
@@ -13,6 +13,8 @@
 using GitHub.Runner.Common;
 using GitHub.Runner.Sdk;
 using GitHub.Services.WebApi.Jwt;
+using System.Text.RegularExpressions;
```
Comment (ericsciple) [inline src/Runner.Listener/JobDispatcher.cs:16]: consider sorting these (i see other system at the top)

**Review by ericsciple — APPROVED** [review]

File: src/Runner.Listener/JobDispatcher.cs (hunk 1 above)
Comment (TingluoHuang) [reply src/Runner.Listener/JobDispatcher.cs]: fixed

File: src/Runner.Worker/Worker.cs [hunk 2]
```diff
@@ -64,7 +66,22 @@ public async Task<int> RunAsync(string pipeIn, string pipeOut)
...
+                    try
+                    {
+                        jobMessage = StringUtil.ConvertFromJson<Pipelines.AgentJobRequestMessage>(channelMessage.Body);
+                    }
+                    catch (JsonReaderException ex)
+                    {
+                        if (channelMessage.Body.Length > ex.LinePosition + 10)
+                        {
+                            var errorChunk = channelMessage.Body.Substring(ex.LinePosition - 10, 20);
+                            terminal.WriteError($"Worker received invalid Json at position '{ex.LinePosition}': {errorChunk} ({Convert.ToBase64String(Encoding.UTF8.GetBytes(errorChunk))})");
```
[the remark this answers is not in the document]
Comment (TingluoHuang) [reply src/Runner.Worker/Worker.cs:79]: the substring is not starts with BOM

File: src/Runner.Listener/JobDispatcher.cs [hunk 3]
```diff
@@ -964,6 +967,30 @@ private async Task LogWorkerProcessUnhandledException(IJobServer jobServer, Pipe
                 TimelineRecord jobRecord = timeline.Records.FirstOrDefault(x => x.Id == message.JobId && x.RecordType == "Job");
                 ArgUtil.NotNull(jobRecord, nameof(jobRecord));
 
+                try
+                {
+                    if (!string.IsNullOrEmpty(errorMessage) &&
+                        message.Variables.TryGetValue("DistributedTask.EnableRunnerIPCDebug", out var enableRunnerIPCDebug) &&
```
Comment (TingluoHuang) [inline src/Runner.Listener/JobDispatcher.cs:973]: Only enabled this trace for customers who have this issue.

**Review by ericsciple — APPROVED** [review]

**Pull request merged**